src/include/libpmemobj++/transaction.hpp

```diff
@@ -223,23 +212,24 @@ class transaction {
 		 * End pmemobj transaction. Depending on the context
 		 * of object destruction, the transaction will
 		 * automatically be either committed or aborted.
+		 *
+		 * @throw nvml::transaction_error if the transaction got aborted
+		 * without an active exception.
 		 */
-		~automatic() noexcept
+		~automatic() noexcept(false)
 		{
-			/* manual abort or commit end transaction */
-			if (pmemobj_tx_stage() != TX_STAGE_WORK) {
-				(void)pmemobj_tx_end();
+			/* active exception, abort handled by tx_worker */
+			if (exceptions.new_uncaught_exception())
 				return;
-			}
 
-			if (this->exceptions.new_uncaught_exception())
-				/* exit with an active exception */
-				pmemobj_tx_abort(ECANCELED);
-			else
-				/* normal exit commit tx */
+			/* transaction ended normally */
+			if (pmemobj_tx_stage() == TX_STAGE_WORK)
 				pmemobj_tx_commit();
-
-			(void)pmemobj_tx_end();
+			/* transaction aborted, throw an exception */
+			else if (pmemobj_tx_stage() == TX_STAGE_ONABORT ||
+				 (pmemobj_tx_stage() == TX_STAGE_FINALLY &&
+				  pmemobj_tx_errno() != 0))
+				throw transaction_error("Transaction aborted");
```

Comment:
else _unreachable();
?

Reply:
There is also TX_STAGE_ONCOMMIT, or not?

Reply:
yup, that should be in the if /w TX_STAGE_WORK.

Reply:
No it shouldn't? It'll be handled by a call to `transaction::~manual` at the end of the scope, hence no else clause. Or am I missing something?